doc/code_coverage_rate/code_coverage_rate.md

```diff
@@ -0,0 +1,173 @@
+![](figure24.png)<br>
+__Figure 2.4 run script in sonic folder__.<br>
+
+![](figure25.png)<br>
+__Figure 2.5 Source file path recorded in info files__.<br>
+
+## 2.4 Check the coverage report
+Figure 2.6 shows that the coverage reports will be saved under /sonic/gcov\_output directory.<br>
+
+![](figure26.png)<br>
+__Figure 2.6 Contents in gcov\_output__.<br>
+
+The output includes:<br>
+info folder: save info files for all modules according to their generated paths<br>
+HTML folder: save gcov HTML report for all modules according to their generated paths<br>
+AllMergeReport: save the merged overall report which contains all coverage information<br>
+info\_err\_list:record the modules which failed to generate the info files<br>
+gcda\_dir\_list.txt:record the directories of all .gcda files<br>
+The merged overall report can be checked from index.html under AllMergeReport dir by a web browser.<br>
+
+![](figure27.png)<br>
+__Figure 2.7 Contents of AllMergeReport__.<br>
+![](figure28.png)<br>
+__Figure 2.8 Overall coverage report__.<br>
+
+# 3. Modules already available for gcov
+  sonic-isc-dhcp<br>
+  sonic-libteam<br>
+  sonic-lldp<br>
+  sonic-frr<br>
+  igb<br>
+  iproute2<br>
+  python3<br>
+  mpdecimal<br>
+  sonic-linux-kernel<br>
+  sonic-platform-modules-ruijie(ruijie-BSP)<br>
+# 4. Safety instruction
+The gcov support for SONiC totally depends on the open-source tools -- gcov/lcov. Hence the modification towards the sonic project is only limited to the compiling options of gcc in order that the additional gcov-required files (.gcno and .gcda) can be generated during compiling. This modification will not have any influence on other sections of the compiling process.<br> 
+
+When a source file is compiled with the gcov compiling options, each executable line in this source file will be followed by a new-added piece of code that updates coverage statistics. Gcov realizes this process by adding stubs when generating assembly files. Each stub point will be inserted into 3 to 4 new assembly statements. These statements are directly added to the .s files. Then the assembly files can be assembled to the object files and the executable file. After doing this, when the executable file is running, the stubs added during compiling will collect the execution information. The statistical approach for these stubs is very simple, they are just variables in the memory and record the execution times for each code line. Therefore in the practical running environment, the performance impact brought by generating .gcda files can be ignored. The user can also hardly feel the difference.<br>
+
+# 5. Coverage data for the python script
```

Comment:
suggest to remove section 5 since it is not gcov related.

Reply:
Ok, done!